### PR TITLE
Optimize context json parsing

### PR DESCRIPTION
**Pull request by benelser**
## Description of changes

This PR removes redundant deserialization when converting an already-parsed `serde_json::Value` into `CedarValueJson`.

Previously, the `from_json_value` path re-entered Serde via `serde_json::from_value`, incurring a full visitor-based deserialization and allocating intermediate structures. This PR introduces a direct conversion path (`CedarValueJson::from_serde_value_direct`) that operates directly on `serde_json::Value`, preserving semantics while avoiding the extra pass.

**Impact:**
- Eliminates double parsing on the `from_json_value` path
- Reduces memory amplification from **7.27× → 1.02×** (≈ **85% reduction**)
- Improves performance and allocation behavior in security- and policy-critical context parsing

Semantic equivalence with the previous implementation was validated via call-graph analysis, error-behavior review, and targeted edge-case checks (including escape handling and error classification).

---

## Issue #, if available

N/A

---

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [x] A change "invisible" to users (e.g., documentation, changes to "internal" crates like `cedar-policy-core`, `cedar-validator`, etc.)

I confirm that this PR (choose one, and delete the other options):

- [x] Does not update the CHANGELOG because this change does not affect public APIs or user-visible behavior.

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [x] Does not require updates because this change does not affect the Cedar formal model or DRT infrastructure.

I confirm that [`docs.cedarpolicy.com`](https://docs.cedarpolicy.com/) (choose one, and delete the other options):

- [x] Does not require updates because this change does not affect the Cedar language specification.

### CODE REVIEW

**Comment by john-h-kastner-aws** [comment]
This optimization makes sense to me and looks promising, but we're very concerned with maintaining the correctness of Cedar and not accidentally introducing any breaking changes. 

Can you please elaborate on what exactly you mean by 

> Semantic equivalence with the previous implementation was validated via call-graph analysis, error-behavior review, and targeted edge-case checks (including escape handling and error classification).

I don't see any additional testing or static analysis included in this PR

Also, you quote some numbers for reduced memory usage, but do you also have any numbers for the speed up (if any) in performance (either in benchmarks or real applications)?  We'd be particularly interested in hearing about applications of Cedar that are running into performance issues.

File: cedar-policy-core/src/entities/json/value.rs [hunk 1]
```diff
@@ -801,7 +927,8 @@ impl<'e> ValueParser<'e> {
                 )),
                 val => {
                     let actual_val = {
-                        let jvalue: CedarValueJson = serde_json::from_value(val)?;
+                        // Use direct conversion for consistency and performance
+                        let jvalue = CedarValueJson::from_serde_value_direct(val, ctx)?;
```
Comment (john-h-kastner-aws) [inline cedar-policy-core/src/entities/json/value.rs:935]: What does "for consistency" mean in this comment? Is it implying there may be a  difference in behavior, or is it simply stating that you use it here because you use it below? 

If the latter, I don't think this comment adds any value

Comment (benelser) [reply cedar-policy-core/src/entities/json/value.rs:935]: Hi John! Yes, it was the latter. I went ahead and removed. Good call out.

**Comment by benelser** [comment]
Thanks @john-h-kastner-aws. I agree that correctness is the primary concern.

I have added explicit tests to back up the “semantic equivalence” claim. The new test module compares the old path (`serde_json::from_value::<CedarValueJson>`) and the new direct path (`from_serde_value_direct`) on a corpus of `serde_json::Value` inputs. It asserts identical `CedarValueJson` results on success, and matches error category on failure. It also includes directed cases for `__entity`, `__extn` (single and multi arg), `__expr` (reserved), and multi key objects containing escape keys to ensure they fall back to `Record`. Null is the one intentional difference at this layer. Serde can produce the `CedarValueJson::Null` variant, while the direct path rejects `Value::Null` immediately. Both code paths still reject null when converting to an expression via `into_expr`, and the tests document that.

On performance, the benchmarks are in `cedar-policy/benches/context_json_parsing.rs`. They measure wall time via Criterion for nested contexts and many attributes, and they include a `from_json_value` benchmark that targets the optimized path specifically. The memory amplification numbers in the PR come from the RSS-based measurements in this benchmark suite.

**Comment by benelser** [comment]
Hey @john-h-kastner-aws, just checking in on this one. I’m curious if the new tests and benchmark data I added covered the concerns you had around potential breaking changes? 

Everything is currently green in CI with solid coverage, but let me know if there's any other specific edge case you're wondering about that I should verify.

**Comment by benelser** [comment]
@john-h-kastner-aws anything else you wanted me to take a look at?

**Comment by victornicolet** [comment]
This PR might be a partial solution to https://github.com/cedar-policy/cedar/issues/2265
